### PR TITLE
mediatek: Add support for D-Link EAGLE PRO AI R32

### PR DESCRIPTION
**Pull request by RolandoMagico**
R32 is like the M32 part of the EAGLE PRO AI series from D-Link.

Specification:
 - MT7622BV SoC with 2.4GHz wifi
 - MT7975AN + MT7915AN for 5GHz
 - MT7531BE Switch
 - 512MB RAM
 - 128 MB flash
 - 2 LEDs (Status and Internet, both can be either orange or white)
 - 2 buttons (WPS and Reset)

Compared to M32, the R32 has the following differences:
 - 4 LAN ports instead of 2
 - The recory image starts with DLK6E6015001 instaed of DLK6E6010001
 - Individual LEDs for power and internet
 - MAC address is stored at another offset in the ODM partition

MAC addresses:
 - WAN MAC is stored in partition "Odm" at offset 0x81
 - LAN (as printed on the device) is WAN MAC + 1
 - WLAN MAC (2.4 GHz) is WAN MAC + 2
 - WLAN MAC (5GHz) is WAN MAC + 3

Flashing via Recovery Web Interface:
 - Set your IP address to 192.168.0.10, subnetmask 255.255.255.0
 - Press the reset button while powering on the deivce
 - Keep the reset button pressed until the internet LED blinks fast
 - Open a Chromium based and goto http://192.168.0.1
 - Download openwrt-mediatek-mt7622-dlink_eagle-pro-ai-r32-a1-squashfs-recovery.bin

Flashing via uBoot:
 - Open the case, connect to the UART console
 - Set your IP address to 10.10.10.3, subnet mask 255.255.255.0. Connect to one of the LAN interfaces of the router
 - Run a tftp server which provides openwrt-mediatek-mt7622-dlink_eagle-pro-ai-r32-initramfs-kernel.bin.
 - You can rename the file to iverson_uImage (no extension), then you don't have to enter the whole file name in uboot later.
 - Power on the device and select "1. System Load Linux to SDRAM via TFTP." in the boot menu
 - Enter image file, tftp server IP and device IP (if they differ from the default).
 - TFTP download to RAM will start. After a few seconds OpenWrt initramfs should start
 - The initramfs is accessible via 192.168.1.1, change your IP address accordingly (or use multiple IP addresses on your interface)
 - Create a backup of the Kernel1 partition, this file is required if a revert to stock should be done later
 - Perform a sysupgrade using openwrt-mediatek-mt7622-dlink_eagle-pro-ai-r32-squashfs-sysupgrade.bin
 - Reboot the device. OpenWrt should start from flash now

Revert back to stock using the Recovery Web Interface:
 - Set your IP address to 192.168.0.10, subnetmask 255.255.255.0
 - Press the reset button while powering on the deivce
 - Keep the reset button pressed until the internet LED blinks fast
 - Open a Chromium based and goto http://192.168.0.1
 - Flash a decrypted firmware image from D-Link. Decrypting an firmware image is described below.

Decrypting a D-Link firmware image:
 - Download https://github.com/RolandoMagico/firmware-utils/blob/M32/src/m32-firmware-util.c
 - Compile a binary from the downloaded file, e.g. gcc m32-firmware-util.c -lcrypto -o m32-firmware-util
 - Run ./m32-firmware-util R32 --DecryptFactoryImage <OriginalFirmware> <OutputFile>
 - Example for firmware R32A1_FW103B01: ./m32-firmware-util R32 --DecryptFactoryImage R32A1_FW103B01.bin R32A1_FW103B01.decrypted.bin

Revert back to stock using uBoot:
 - Open the case, connect to the UART console
 - Set your IP address to 10.10.10.3, subnet mask 255.255.255.0. Connect to one of the LAN interfaces of the router
 - Run a tftp server which provides the previously created backup of the Kernel1 partition.
 - You can rename the file to iverson_uImage (no extension), then you don't have to enter the whole file name in uboot later.
 - Power on the device and select "2. System Load Linux Kernel then write to Flash via TFTP." in the boot menu
 - Enter image file, tftp server IP and device IP (if they differ from the default).
 - TFTP download to FLASH will start. After a few seconds the stock firmware should start again

There is also an image openwrt-mediatek-mt7622-dlink_eagle-pro-ai-r32-a1-squashfs-tftp.bin which can directly be flashed via U-Boot and TFTP. It can be used if no backup of the Kernel1 partition is reuqired.

Flahsing via OEM web interface is currently not possible, the OEM images are encrypted. Creating images is only possible manually at the moment.
The support for the M32/R32 already includes support for flashing from the OEM web interface:
 - The device tree contains both partitions (Kernel1 and Kernel2) with conditions to select the correct one based on the kernel command line
 - The U-Boot variable "boot_part" is set accordingly during startup to finish the partition swap after flashing from the OEM web interface
 - OpenWrt sysupgrade flashing always uses the partition where it was initially flashed to (no partition swap)

### CODE REVIEW

File: target/linux/mediatek/mt7622/base-files/etc/board.d/02_network
Comment (musashino205) [inline target/linux/mediatek/mt7622/base-files/etc/board.d/02_network:32]: same as ```bananapi,bpi-r64|\ ... ruijie,rg-ew3200gx-pro)``` section

Comment (RolandoMagico) [reply target/linux/mediatek/mt7622/base-files/etc/board.d/02_network]: Fixed

**Comment by RolandoMagico** [comment]
I updated to the lastest main which includes setting the MACs in the device tree instead of using the following files:
- /etc/board.d/02_network
- /etc/hotplug.d/iee80211/11_fix_wifi_mac

Looks like incrementing the MAC adresses within `nvmem-cells = <macaddr_odm, x>` doesn't work so the device tree still contains `mac-address-increment = <x>`.

@neheb, @Ansuel : Any idea how to fix this?

**Comment by neheb** [comment]
I see no diff.

**Comment by RolandoMagico** [comment]
> I see no diff.

This are the changes for wifi: https://github.com/openwrt/openwrt/compare/2ac58a7e5511405826149271aefc46167798353f..6c43b07e35de36962befded713bf8646ab7d523f

The wifi mac adresses (macaddr_odm + 2 and macaddr_odm +3) are only working if mac-address-increment is present, otherwise the value in macaddr_odm is used without adding a value.

**Comment by neheb** [comment]
I believe it doesn’t work for UBI yet. In my conversions, I’ve avoided touching user space MAC stuff.

**Comment by neheb** [comment]
@RolandoMagico btw macaddr_odm needs to use mac-base.

**Comment by RolandoMagico** [comment]
Thanks @neheb, now it's working as expected. The MAC addresses are set correctly without having to use `mac-address-increment` anymore.
Regarding mac-base: Do you mean `compatible = "mac-base";`? If yes: This is already present in the device specific DTS files (I had to put it there because the MAC adress is stored at different offsets for M32 and R32).

File: target/linux/mediatek/image/mt7622.mk
Comment (raenye) [inline target/linux/mediatek/image/mt7622.mk:54]: is this comment still needed?

File: target/linux/mediatek/dts/mt7622-dlink-eagle-pro-ai-ax3200-a1.dtsi
Comment (RolandoMagico) [reply target/linux/mediatek/dts/mt7622-dlink-eagle-pro-ai-ax3200-a1.dtsi]: Changed

File: target/linux/mediatek/dts/mt7622-dlink-eagle-pro-ai-ax3200-a1.dtsi
Comment (RolandoMagico) [reply target/linux/mediatek/dts/mt7622-dlink-eagle-pro-ai-ax3200-a1.dtsi]: Changed

File: target/linux/mediatek/image/mt7622.mk
Comment (RolandoMagico) [reply target/linux/mediatek/image/mt7622.mk]: Doesn't make sense it the way it currently is. Tried to make something more useful out of it.

Comment (raenye) [reply target/linux/mediatek/image/mt7622.mk]: @RolandoMagico 
I thought it makes sense to compute the checksum in the makefile, but something doesn't work for me:
```sh
header='DLK6E6010001\x00\x00\xCF\x33\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x8D\x57\x30\x0B\x00\x00\x2C\x00\x00\x00\xD0\x02\x00\x00\x2C\x00\x00\x00\xD0\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x42\x48\x02\x00\x00\x00\x08\x00\x00\x00\x00\x00\x60\x6E\x68\x61'
echo -ne $header | od -v -w2 -tu2 -An --endian little | paste -sd+ | bc | xargs printf '%04x'
```
prints `2fff4`, not `0000`. Same result for the other model.
I tried big endian instead, still no go.

Comment (raenye) [reply target/linux/mediatek/image/mt7622.mk]: So it's sum mod 65535. not mod 65536. How about this:
```sh
# create $@.header without the checksum
od -v -w2 -tu2 -An --endian little $@.header | awk '{s+=$1;}END{s=65535-s%65535;printf "%c%c",s%256,s/256;}' >> $@.header
```

Comment (RolandoMagico) [reply target/linux/mediatek/image/mt7622.mk]: I tried it:
```
# Only zeros
	echo -en "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00" >> "$@.header"
# Last 16 bytes, but without checksum
	echo -en "\x42\x48\x02\x00\x00\x00\x08\x00\x00\x00\x00\x00\x60\x6E" >> "$@.header"
# Calculate and append checksum
	od -v -w2 -tu2 -An --endian little $@.header | awk '{s+=65535-$1;}END{s%=65535;printf "%c%c",s%256,s/256;}' > $@.header.checksum
	cat "$@.header" "$@.header.checksum" "$@" > "$@.new"
	mv "$@.new" "$@"
	rm "$@.header" "$@.header.checksum"
```
But checksum is 00 00 for both devices.

Comment (raenye) [reply target/linux/mediatek/image/mt7622.mk]: You don't want `s = 65535 - s` since checksum 0 needs to remain 0. It can be `s = (65535 - s) % 65535` though.
And if you insist on doing the modulo every time, `s+=$1; s%=65535;` makes more sense.

How about
```sh
# Calculate and append checksum
( \
	cat "$@.header"; \
	od -v -w2 -tu2 -An --endian little $@.header | awk '{s+=65535-$1;}END{s%=65535;printf "%c%c",s%256,s/256;}'; \
	cat "$@"; \
) > "$@.new"
```

Comment (raenye) [reply target/linux/mediatek/image/mt7622.mk]: > I want to simulate 16bit overflows using the `s%=65536`

If you want it every numberyou could do either
```sh
awk '{s+=$1;s%=65535;}END{s=65535-s;printf "%c%c",s%256,s/256;}'
```
or
```sh
awk '{s+=65535-$1;s%=65535;}END{printf "%c%c",s%256,s/256;}'
```
My version does it once, at the end.

Comment (RolandoMagico) [reply target/linux/mediatek/image/mt7622.mk]: Ok, I think now I understood your script. All three versions from you are working on command line printing "ha" (0x68 0x61) for the M32, but if I'm using it in the Makefile, I get only 00 00 in the binary. I also see that the file $@.header.checksum already contains the incorrect values.

Comment (RolandoMagico) [reply target/linux/mediatek/image/mt7622.mk]: Escaping $ in $1 to $$1 (Makefiles...) was missing

**Comment by RolandoMagico** [comment]
> May I also nag about LEDs? (color and function properties instead of label, node names `led-xxx`)

Sure, do you have a recommendation?

**Comment by RolandoMagico** [comment]
> See, e.g., [e29f4a3](https://github.com/openwrt/openwrt/commit/e29f4a3f70d278cf7e360519d97d94f54baae278).

Do I have to know anything special about the changed LED configuration? Currently it looks like the LEDs change their behavior: The usual red/orange blinking during boot and firmware upgrade is missing and the white power LED isn't on after boot.

As soon as I assign labels to the led entries in the DTS, the usual behavior is back.

**Comment by RolandoMagico** [comment]
> Until #13837 is merged, LEDs referenced in aliases need to have a `label`. (I cherry picked this for my local tree).

Understood. Can cherry-picking be modelled in a pull request? If not: I subscribed to https://github.com/openwrt/openwrt/pull/13837, so I'll do the remaining changes as soon as it is merged.

File: target/linux/mediatek/dts/mt7622-dlink-eagle-pro-ai-ax3200-a1.dtsi
Comment (musashino205) [inline target/linux/mediatek/dts/mt7622-dlink-eagle-pro-ai-ax3200-a1.dtsi:129]: broken indent

Comment (RolandoMagico) [reply target/linux/mediatek/dts/mt7622-dlink-eagle-pro-ai-ax3200-a1.dtsi]: Thanks, fixed

**Comment by raenye** [comment]
> Understood. Can cherry-picking be modelled in a pull request?

AFAIK no. In my local git tree I have per-PR branches, which only have the commits I want to push, and a `mybuild` branch which has everything including some not yet merged commits from PRs I like.

**Comment by RolandoMagico** [comment]
> > I'll do the remaining changes as soon as it is merged.
> 
> It has just been merged :)

Done

**Comment by s-2** [comment]
rebuilt from your branch, sysupgraded my device (settings erased) from whatever was running on it before... needed a power cycle before ethernet would work, not sure if this is intended (due to bootloader or something? WAN LED was glowing both white and orange). 
After power cycle, both wifi interfaces needed to be enabled before scan was possible, but I think we've had that before.
Otherwise, everything is working fine, thanks for the PR :slightly_smiling_face:

**Comment by RolandoMagico** [comment]
> both wifi interfaces needed to be enabled before scan was possible, but I think we've had that before.

I saw the issue also on other devices and opened https://github.com/openwrt/openwrt/issues/14207

**Comment by RolandoMagico** [comment]
Is there anything to do from my side to get the PR accepted? Or do I just have to wait for the required approval?

**Comment by neheb** [comment]
wait unfortunately.

**Review by neheb — APPROVED** [review]

**Comment by neheb** [comment]
ping @dangowrt

File: target/linux/mediatek/image/mt7622.mk
Comment (dangowrt) [inline target/linux/mediatek/image/mt7622.mk:202]: No `tftp.bin` for the r32 like we got for the m32?

I kinda find the `recovery.bin` naming inconsistent with the existing use of that image name suffix (which hints that the file is used *as* recovery image rather than to be flashed *by* a vendor's recovery system), but that already happened now...

Comment (RolandoMagico) [reply target/linux/mediatek/image/mt7622.mk:202]: The tftp image is created for both devices. It's defined in Device/dlink_eagle-pro-ai-ax3200-a1 and used by Device/dlink_eagle-pro-ai-m32-a1 and Device/dlink_eagle-pro-ai-r32-a1.
Any suggestions for the naming of the recovery image? I took over the naming from D-Link DAP-X1860.

Comment (raenye) [reply target/linux/mediatek/image/mt7622.mk:202]: Wouldn't `initramfs.bin` be considered the recovery for OpenWrt?
AFAIK `recovery.bin` is for vendor uboot recovery in several devices already.

**Comment by RolandoMagico** [comment]
> needed a power cycle before ethernet would work, not sure if this is intended (due to bootloader or something? WAN LED was glowing both white and orange).

I cannot reproduce it with latest main. If you want to retry it, that's the version I used:
https://github.com/RolandoMagico/openwrt-build/releases/tag/EAGLE_PRO_AI_R32_Build_20231218

**Comment by RolandoMagico** [comment]
I updated the recovery image creation because during porting the M30, it turned out that there is another checksum in the header of the recovery image. It's currently not used by the recovery web interface, but if it would be considered in newer devices, it would work as well.

**Pull request merged**
